sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/UnsafeArraySuite.scala

```diff
@@ -18,27 +18,126 @@
+      toRow(floatArray).getArray(0)
+    assert(unsafeFloat.isInstanceOf[UnsafeArrayData])
+    assert(unsafeFloat.numElements == floatArray.length)
+    floatArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeFloat.getFloat(i) == e)
+    }
+
+    val unsafeDouble = ExpressionEncoder[Array[Double]].resolveAndBind().
+      toRow(doubleArray).getArray(0)
+    assert(unsafeDouble.isInstanceOf[UnsafeArrayData])
+    assert(unsafeDouble.numElements == doubleArray.length)
+    doubleArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeDouble.getDouble(i) == e)
+    }
+
+    val unsafeMultiDimInt = ExpressionEncoder[Array[Array[Int]]].resolveAndBind().
+      toRow(intMultiDimArray).getArray(0)
+    assert(unsafeMultiDimInt.isInstanceOf[UnsafeArrayData])
+    assert(unsafeMultiDimInt.numElements == intMultiDimArray.length)
+    intMultiDimArray.zipWithIndex.map { case (a, j) =>
+      val u = unsafeMultiDimInt.getArray(j)
+      assert(u.isInstanceOf[UnsafeArrayData])
+      assert(u.numElements == a.length)
+      a.zipWithIndex.map { case (e, i) =>
+        assert(u.getInt(i) == e)
+      }
+    }
+
+    val unsafeMultiDimDouble = ExpressionEncoder[Array[Array[Double]]].resolveAndBind().
+      toRow(doubleMultiDimArray).getArray(0)
+    assert(unsafeDouble.isInstanceOf[UnsafeArrayData])
+    assert(unsafeMultiDimDouble.numElements == doubleMultiDimArray.length)
+    doubleMultiDimArray.zipWithIndex.map { case (a, j) =>
+      val u = unsafeMultiDimDouble.getArray(j)
+      assert(u.isInstanceOf[UnsafeArrayData])
+      assert(u.numElements == a.length)
+      a.zipWithIndex.map { case (e, i) =>
+        assert(u.getDouble(i) == e)
+      }
+    }
+  }
+
+  test("from primitive array") {
+    val unsafeInt = UnsafeArrayData.fromPrimitiveArray(intArray)
+    assert(unsafeInt.numElements == 3)
+    assert(unsafeInt.getSizeInBytes ==
+      ((4 + scala.math.ceil(3/64.toDouble) * 8 + 4 * 3 + 7).toInt / 8) * 8)
+    intArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeInt.getInt(i) == e)
+    }
+
+    val unsafeDouble = UnsafeArrayData.fromPrimitiveArray(doubleArray)
+    assert(unsafeDouble.numElements == 3)
+    assert(unsafeDouble.getSizeInBytes ==
+      ((4 + scala.math.ceil(3/64.toDouble) * 8 + 8 * 3 + 7).toInt / 8) * 8)
+    doubleArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeDouble.getDouble(i) == e)
+    }
   }
 
-  test("from primitive double array") {
-    val array = Array(1.1, 2.2, 3.3)
-    val unsafe = UnsafeArrayData.fromPrimitiveArray(array)
-    assert(unsafe.numElements == 3)
-    assert(unsafe.getSizeInBytes == 4 + 4 * 3 + 8 * 3)
-    assert(unsafe.getDouble(0) == 1.1)
-    assert(unsafe.getDouble(1) == 2.2)
-    assert(unsafe.getDouble(2) == 3.3)
+  test("to primitive array") {
+    val intEncoder = ExpressionEncoder[Array[Int]].resolveAndBind()
+    val intInternalRow = intEncoder.toRow(intArray)
+    val intUnsafeArray = intInternalRow.getArray(0)
```

Comment:
we can combine these 2 lines: `val unsafeIntArray = intEncoder.toRow(intArray).getArray(0)`